src/Microsoft.VisualStudio.Composition/Configuration/Formatters/MetadataObjectFormatter.cs

```diff
@@ -151,45 +153,80 @@ public void Serialize(ref MessagePackWriter messagePackWriter, object? value, Me
         }
     }
 
-    public object? Deserialize(ref MessagePackReader reader, MessagePackSerializerOptions options)
+    public ImmutableDictionary<string, object?> DeserializeMetadataObjects(ref MessagePackReader reader, MessagePackSerializerOptions options, int count)
     {
-        if (reader.TryReadNil())
+        var dictionary = new Dictionary<string, object?>(count);
+
+        try
+        {
+            var stringFormatter = options.Resolver.GetFormatterWithVerify<string>();
+            var typeRefFormatter = options.Resolver.GetFormatterWithVerify<TypeRef?>();
+            var guidFormatter = options.Resolver.GetFormatterWithVerify<Guid>();
+            var typeRefFormatterCollection = options.Resolver.GetFormatterWithVerify<IReadOnlyList<TypeRef>>();
+
+            for (int i = 0; i < count; i++)
+            {
+                dictionary.Add(stringFormatter.Deserialize(ref reader, options), this.Deserialize(ref reader, options, stringFormatter, typeRefFormatter, guidFormatter, typeRefFormatterCollection));
+            }
+        }
+        finally
         {
-            return null;
+            reader.Depth--;
```

Comment:
👮🏼 Oops. You're decrementing where you didn't increment. This line is missing above the `try` block:

```cs
options.Security.DepthStep(ref messagePackReader);
```

Apparently the code you moved this from had this bug too.

Reply:
Only in case of loop

Reply:
check at allother places